modules/nextflow/src/main/groovy/nextflow/splitter/CsvSplitter.groovy

```diff
@@ -167,13 +185,65 @@ class CsvSplitter extends AbstractTextSplitter {
         if( !columnsHeader )
             return tokens
 
+        if( columnTypes && columnTypes.size() != tokens.size() )
+            throw new IllegalArgumentException("The number of column types should match the number of csv columns. Provided types: ${columnTypes}")
+
         def map = [:]
-        for( int i=0; i<columnsHeader.size(); i++ )
-            map[ columnsHeader[i] ] = i<tokens.size() ? tokens[i] : null
+        for( int i = 0; i < tokens.size(); i++ ) {
+            if( typesAuto )
+                map[columnsHeader[i]] = inferValueType(tokens[i])
```

Comment:
I would like to use this functionality in the nf-validation pluging, where I validate a samplesheet csv file automatically, so the user won't be able to provide the specific types, which makes it difficult to validate the file without the automatic detection of variable types.

If maintaining the automatic type detection is not feasible, implementing it directly in the nf-validation plugin could be the best solution. Do you think this would be the best approach?

Reply:
But don't we have an explicit schema for the sample sheet? So we should know the types?

Reply:
Paolo, I think we might as well allow the auto type inference and say "use at your own risk". Users can always fall back to explicit types.

It's a common feature in CSV parsers, for example see [pandas.read_csv](https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html). So if you're really concerned about it, we could even see what these established implementations are doing and try to imitate it.

Reply:
> But don't we have an explicit schema for the sample sheet? So we should know the types?

Right, I didn't think about that. We currently only use the schema after parsing the CSV, but we could obtain the types before. The only thing I'm not sure about is the order of properties in the schema, but I can probably solve this with the column names. I'll try it out. Thanks!

Reply:
> Paolo, I think we might as well allow the auto type inference and say "use at your own risk". Users can always fall back to explicit types

the reality people don't care and just use the quickest way, when you will discovered it will be too late